### ISSUE TITLE
Too many PKSim.CLI processes

### ISSUE THREAD

**Issue by Yuri05** [Open-Systems-Pharmacology/QualificationRunner#108]
When I call the qualification runner from the reporting engine with a bigger qualification plan: the number of PKSim.CLI processes launched in parallel exceeds the number of cores. 
E.g. I have 8 cores and when launching the CYP3A4-DDI qualification plan: 20 and sometimes even more PKSim.CLI are executed in parallel.
![grafik](https://user-images.githubusercontent.com/25061876/154994977-0b73efc8-e1de-41d2-af01-3c671d5df4df.png)

**Comment by msevestre** [comment]
It is this task
https://github.com/Open-Systems-Pharmacology/QualificationRunner/blob/develop/src/QualificationRunner.Core/Services/QualificationRunner.cs#L76

Is this bad? the system should be able to deal with this. Or what is your suggestion? one at a time?

**Comment by Yuri05** [comment]
Not one at a time of course but the possibility to restrict the number of PKSim.CLI started in parallel.
It's not bad I think (in terms of performance), the only point is that currently almost 100% CPU is consumed and users asked already in the past how the CPU usage of the quali runner can be restricted.

**Comment by msevestre** [comment]
I am not quite sure how to limit this. Do you want a parameter e.g. max number of cores or sthg like this. We'll need to rewrite the loop above also

**Comment by Yuri05** [comment]
Yes, it would be nice to have an optional parameter `Number of cores ` which is passed to the Qualification Runner. 
If not set: use `MaxCores-1 `

**Comment by msevestre** [comment]
Nice as in...do it for v11 or when we have time ...

**Comment by rwmcintosh** [comment]
That should fit in V13 nicely.

**Comment by rwmcintosh** [comment]
> When I call the qualification runner from the reporting engine with a bigger qualification plan: the number of PKSim.CLI processes launched in parallel exceeds the number of cores. E.g. I have 8 cores and when launching the CYP3A4-DDI qualification plan: 20 and sometimes even more PKSim.CLI are executed in parallel. ![grafik](https://user-images.githubusercontent.com/25061876/154994977-0b73efc8-e1de-41d2-af01-3c671d5df4df.png)

But this looks like separate processes actually, not threads within a process.

Edit++++++++

NVM, I see that this ultimately uses a Task.Run to start a new process. That explains it.

**Comment by rwmcintosh** [comment]
Maybe we can adjust the process priority instead. They can all start at once and if the user gives a priority flag then we can run the new processes with lower priority.

**Comment by msevestre** [comment]
This is because we run a new Task all the time. How do you want to change this? aren't they differnet process completely ?

**Comment by msevestre** [comment]
>Yes, it would be nice to have an optional parameter Number of cores  which is passed to the Qualification Runner.
If not set: use MaxCores-1 

what about this?

**Comment by rwmcintosh** [comment]
> > Yes, it would be nice to have an optional parameter Number of cores  which is passed to the Qualification Runner.
> > If not set: use MaxCores-1
> 
> what about this?

I'm just not sure how the Task.WhenAll relates to the startable process. Do they wait for the process to complete?

If the process priority works, that would be less work because we could leave the loop as it is. with Task.WhenAll

**Comment by rwmcintosh** [comment]
yes, there's a wait in there `process.Wait(cancellationToken);`

**Comment by rwmcintosh** [comment]
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.priorityclass?view=net-8.0

**Comment by rwmcintosh** [comment]
> This is because we run a new Task all the time. How do you want to change this? aren't they differnet process completely ?

Yeah, if we give all new processes a `BelowNormal` priority, then we should always pre-empt them with any GUI application.
